### ISSUE TITLE
Agent status issues

### ISSUE THREAD

**Issue by agentsofshield** [wazuh/wazuh#3317]
Wazuh version: 3.9

Hello, I've ran into several problems.

I use the splunk app and I recently noticed that there are too many disconnected agents (7000 out of 10000), furthermore the agents status on the servers is the same, then I saw that the agent on my own computer is also shown as disconnected even though it is working (sends logs, syscheck, rootcheck...).
Another thing I noticed is that the agent version reported is not the actual agent version, it is the older version. 
Lastly I noticed that the last keep alive time and last syscheck scan time is in the **future**. The time on the servers and on my computer is exactly the same.

I can't interact with the disconnected agents from splunk because they are disconnected even though I know they work fine.

Thanks in advance!

**Comment by adri9valle** [comment]
Hi @agentsofshield,

In order to determine where the problem is coming from, please follow these steps:

We are going to ensure that the Wazuh is working properly. The first please execute the following commands in order to get the number of active and disconnected agents:


- The number of active agents:

```
/var/ossec/bin/agent_control -l | fgrep Active | wc -l
```    

- The number of disconnected agents:

```
/var/ossec/bin/agent_control -l | fgrep Disconnected | wc -l
```

The previous commando will return a number, now let's see if these numbers match with the info that the API returns. Please in the manager/master instance execute the below command:

```
curl -u {user}:{pass} http://localhost:55000/agents/summary?pretty
```    

The API response should be something like this:

```
{
   "error": 0,
   "data": {
      "Total": A,
      "Active": B,
      "Disconnected": C,
      "Never connected": D,
      "Pending": E
   }
}
```


If the results match, then it seems that the API is returning properly the number of the total active and disconnected agents.

You can check a specific agent executing the next command in the manager:

NOTE: We use the agent id 007 for this example.

```
# /var/ossec/bin/agent_control -l | fgrep 007
   ID: 007, Name: agcen2, IP: any, Active
```

Now we gonna check the API request:

```
# curl -u foo:bar http://localhost:55000/agents/007?pretty
{
   "error": 0,
   "data": {
      "os": {
         "arch": "x86_64",
         "major": "7",
         "minor": "5",
         "name": "CentOS Linux",
         "platform": "centos",
         "uname": "Linux |ag-2 |3.10.0-862.14.4.el7.x86_64 |#1 SMP Wed Sep 26 15:12:11 UTC 2018 |x86_64",
         "version": "7.5"
      },
      "status": "Active",
      "mergedSum": "21c0d311586c307f210589388688e3b6",
      "manager": "master",
      "dateAdd": "2019-05-15 09:45:24",
      "name": "agcen2",
      "group": [
         "default"
      ],
      "ip": "10.0.2.15",
      "version": "Wazuh v3.9.0",
      "configSum": "015ce99af52bb4e42764044a7c7b24c6",
      "registerIP": "any",
      "id": "007",
      "node_name": "node01",
      "lastKeepAlive": "2019-05-15 10:36:28"
   }
}
```

We can see the agents status, the last keep alive and the Wazuh version among other things:

```
"status": "Active"
"version": "Wazuh v3.9.0"
"lastKeepAlive": "2019-05-15 10:36:28"
```


Also, we can check the wazuh version, for check the Wazuh version in the agent, please check this file:

```
# cat /etc/ossec-init.conf
DIRECTORY="/var/ossec"
NAME="Wazuh"
VERSION="v3.9.0"
REVISION="3918"
DATE="Fri May  3 01:47:44 UTC 2019"
TYPE="agent"
```

And now in the manager we gonna check the Wazuh version storaged in the **global.db**:


```    
# sqlite3 /var/ossec/var/db/global.db
``` 

```    
sqlite>  select id ,name, version, last_keepalive from agent where id=7;
   
   
7|agcen2|Wazuh v3.9.0|2019-05-15 10:18:48
```    

If the Wazuh version doesn't match then there's a problem when Wazuh updates the database.

Please let me know the results of these tests, this could determinate if the problem comes from the Wazuh core, API or Splunk App.

Regards,

**Comment by agentsofshield** [comment]
Thank you!
I did as you told me and here are the results:

- The numbers I got from the api do match the ones I got directly from the server.
- The last keep alive in the db is correct, a few hours is added when it gets to splunk.
- The wazuh version got updated by itself but it took a lot of time.

I have a load balancer before my cluster, maybe it poses a problem with the agents status?

**Comment by adri9valle** [comment]
Hi @agentsofshield,

Regarding the few hours added to the last keepAlive in Splunk, this is because the App is applying the timezone of the browser, please check if the timezone of the browser and the manager is not the same if is not the same then this is normal.

Related the others problem with the agent synchronization problem, it seems that is for this problem: [The cluster is not synchronizing the agent-info files properly](https://github.com/wazuh/wazuh/issues/3271). This problem was solved a few days ago and will be implemented in Wazuh 3.9.1, you can see the solution here: [Fix aborting loop while reading agent info files](https://github.com/wazuh/wazuh/pull/3272).

That [return](https://github.com/wazuh/wazuh/blob/69729679efce4aed0029ae37f80610aef574f5e8/framework/wazuh/cluster/master.py#L396) breaks the for loop preventing the following agent-info files to be processed. So, the master will never have all the agent-info files (if the condition is met - the agent-info processed is older than the agent-info in the master) and the App will show a wrong number of active/disconnected agents.

You could modify manually this script and change `return` for `continue` until Wazuh 3.9.1 will be released.

I hope this helps you. Please let me know if this is useful to you.

Regards,

**Comment by agentsofshield** [comment]
Thank you very much, everything works now. 
It seems that the problem with the agents was because the load balancer was misconfigured, it sent data to a server which was down so the agents didn't get a reply and were labeled as disconnected or never connected.
